File: Lab 12/History.cs
public class History
{

    private readonly List<EditorMemento> _mementos;
    private readonly TextEditor _texteditor;
    public History(TextEditor textEditor)
    {
        _textEditor = textEditor;
    }
    public void Backup()
    {
        _mementos.Add(_textEditor.CreateMemento());
    }
    public void Undo()
    {
        if (_mementos.Count == 0)
        {
            Console.WriteLine("No mementos to undo.");
            return;
        }
        var lastmemento = _mementos.Last();
        _mementos.Remove(lastmemento);
        _textEditor.RestoreFromMemento(memento);
    }
    public void SaveToFile(string filePath)
    {
        try
        {
            string json = JsonSerializer.Serialize(_mementos);
            File.WriteAllText(filePath, json);
            Console.WriteLine(" History saved to file.");

        }
        catch
         {
            Console.WriteLine("Failed to save history to file.");
         }
    }
    public void LoadFromFile(string filePath)
    {
         try
        {
            string json = File.ReadAllText(filePath) ;
            _mementos.Clear();
            _mementos.AddRange(JsonSerializer.Deserialize<List<EditorMemento>>(json));
            Console.WriteLine(" History loaded from file.");

        }
        catch
         {
            Console.WriteLine("Error loading file");
         }

    }
}

File: Lab 12/src/History.java
import java.util.Stack;

public class History {
    private final Stack<EditorMemento> mementoStack;

    public History() {
        mementoStack = new Stack<>();
    }

    public void backup(TextEditor editor) {
        mementoStack.push(editor.createMemento());
    }

    public void undo(TextEditor editor) {
        if (!mementoStack.isEmpty()) {
            EditorMemento memento = mementoStack.pop();
            editor.restoreFromMemento(memento);
        } else {
            System.out.println("No history to undo.");
        }
    }
}

